File: eval/src/lib/fees.js
// Kalshi fee schedule:
//   taker fee = ceil(0.07   * C * P * (1 - P))  in cents
//   maker fee = ceil(0.0175 * C * P * (1 - P))  in cents
// where C = contracts, P = price in dollars (0..1). Rounded UP to the next cent.
const TAKER = 0.07;
const MAKER = 0.0175;

function feeForFill(qty, priceCents, orderType) {
  if (qty <= 0) return 0;
  const p = priceCents / 100;
  const coef = orderType === "Maker" ? MAKER : TAKER;
  const raw = coef * qty * p * (1 - p);
  // Kalshi rounds UP to the next whole cent. Guard against FP dust.
  const cents = Math.ceil(raw * 100 - 1e-9);
  return cents / 100;
}

// Recompute total fees for a play if every fill's qty were scaled by `scale`.
// Rounding is applied per-fill, matching Kalshi's per-order-book behavior.
export function scaledFees(play, scale) {
  let total = 0;
  for (const f of play.entryFills || []) {
    total += feeForFill(f.qty * scale, f.price, f.orderType);
  }
  for (const f of play.exitFills || []) {
    total += feeForFill(f.qty * scale, f.price, f.orderType);
  }
  return total;
}

// Gross scales exactly linearly (price × qty), so no formula adjustment needed.
export function scaledGross(play, scale) {
  return play.gross * scale;
}

File: backend/app/fees.py
"""
Kalshi fee schedule — single source of truth.

  taker fee = ⌈0.07   × C × P × (1 − P)⌉   (cents)
  maker fee = ⌈0.0175 × C × P × (1 − P)⌉   (cents)

where C is the number of contracts and P is the contract price in dollars
(0..1). Rounded UP to the next whole cent.

JS mirror: eval/src/lib/fees.js.
"""

import math

TAKER_COEF = 0.07
MAKER_COEF = 0.0175


def fee_for_contracts(contracts: float, price_dollars: float, coef: float) -> float:
    """⌈coef × C × P × (1 − P)⌉ rounded UP to the next cent. Returns dollars.

    `contracts` is a count, not a dollar amount.
    """
    if contracts <= 0 or price_dollars <= 0 or price_dollars >= 1:
        return 0.0
    raw_dollars = coef * contracts * price_dollars * (1.0 - price_dollars)
    # Ceiling to the next cent; epsilon guard against FP dust that would
    # otherwise push an exact-cent value up one cent.
    cents = math.ceil(raw_dollars * 100 - 1e-9)
    return cents / 100


def taker_fee(contracts: float, price_dollars: float) -> float:
    return fee_for_contracts(contracts, price_dollars, TAKER_COEF)


def maker_fee(contracts: float, price_dollars: float) -> float:
    return fee_for_contracts(contracts, price_dollars, MAKER_COEF)
